File: Novomber 2022/Day 1/C++ Code/Median in a row-wise sorted Matrix.cpp
// Problem Link :- https://practice.geeksforgeeks.org/problems/median-in-a-row-wise-sorted-matrix1527/1
// Video Solution Link :- https://www.youtube.com/watch?v=-gI6II7eTfc

//{ Driver Code Starts
//Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function template for C++

class Solution{   
public:
    
    // Time Complexity :- O(32.r.log(c));
    // Space Complexity :- O(1);
    
    int median(vector<vector<int>> &matrix, int R, int C){
        // code here  
        int lo = INT_MAX,hi=INT_MIN;
        
        for(int i=0;i<R;i++) {
            lo = min(lo,matrix[i][0]);
            hi = max(hi,matrix[i][C-1]);
        }
        int required_value=((R*C)/2);
        if(R%2==0 || C%2==0) required_value--;
        int ans=hi;
        while(lo<=hi) {
            int mid = lo+(hi-lo)/2;
            
            int count=0;
            for(int i=0;i<R;i++) {
                count+=search(matrix[i],mid);
            }
            
            if(count<=required_value) {
                lo=mid+1;
            } else {
                hi=mid-1;
                ans=min(ans,mid);
            }
        }
        return ans;
    }
    
    int search(vector<int> a,int v) {
        int lo=0,hi=a.size()-1;
        int ans=-1;
        while(lo<=hi) {
            int mid = lo+(hi-lo)/2;
            if(a[mid]<=v) {
                lo=mid+1;
                ans=max(ans,mid);
            } else {
                hi=mid-1;
            }
        }
        return ans+1;
    }
};


//{ Driver Code Starts.

int main()
{
    int t;
    cin>>t;
    while(t--)
    {
        int r, c;
        cin>>r>>c;
        vector<vector<int>> matrix(r, vector<int>(c));
        for(int i = 0; i < r; ++i)
            for(int j = 0;j < c; ++j)
                cin>>matrix[i][j];
        Solution obj;
        cout<<obj.median(matrix, r, c)<<endl;        
    }
    return 0;
}
// } Driver Code Ends

File: Novomber 2022/Day 1/JAVA Code/Median in a row-wise sorted Matrix.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;


class GFG
{
    public static void main(String args[])throws IOException
    {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while(t-- > 0)
        {
            String S[] = read.readLine().split(" ");
            int R = Integer.parseInt(S[0]);
            int C = Integer.parseInt(S[1]);
            int matrix[][] = new int[R][C];
            int c = 0;
            for(int i = 0; i < R; i++){
                String line[]=read.readLine().trim().split(" ");
                for(int j = 0; j < C; j++){
                    matrix[i][j] = Integer.parseInt(line[j]);
                }
            }
            Solution ob = new Solution();
            int ans = ob.median(matrix, R, C);
            System.out.println(ans);
        }
    }
}

// } Driver Code Ends


//User function Template for Java

class Solution {
    
    // Time Complexity :- O(32.r.log(c));
    // Space Complexity :- O(1);
    
    int median(int matrix[][], int R, int C) {
        // code here   
        
        // Brute Force
        // Time Complexity :- O(r.c);
        // Space Complexity :- O(r.c);
        // ArrayList<Integer> list = new ArrayList<>();
        // for(int[] val : matrix) {
        //     for(int v1 : val) {
        //         list.add(v1);
        //     }
        // }
        // Collections.sort(list);
        // int idx = (R*C)/2;
        // if(R%2==0 || C%2==0) idx--;
        // return list.get(idx);
        
        int lo = Integer.MAX_VALUE,hi=Integer.MIN_VALUE;
        
        for(int i=0;i<R;i++) {
            lo = Math.min(lo,matrix[i][0]);
            hi = Math.max(hi,matrix[i][C-1]);
        }
        int required_value=((R*C)/2);
        if(R%2==0 || C%2==0) required_value--;
        int ans=hi;
        while(lo<=hi) {
            int mid = lo+(hi-lo)/2;
            
            int count=0;
            for(int i=0;i<R;i++) {
                count+=search(matrix[i],mid);
            }
            
            if(count<=required_value) {
                lo=mid+1;
            } else {
                hi=mid-1;
                ans=Math.min(ans,mid);
            }
        }
        return ans;
        
    }
    
    public static int search(int[] a,int v) {
        int lo=0,hi=a.length-1;
        int ans=-1;
        while(lo<=hi) {
            int mid = lo+(hi-lo)/2;
            if(a[mid]<=v) {
                lo=mid+1;
                ans=Math.max(ans,mid);
            } else {
                hi=mid-1;
            }
        }
        return ans+1;
    }
}
